contracts/zaps/ZapTrancheLp.sol

```diff
@@ -0,0 +1,277 @@
+            msg.sender,
+            address(this),
+            ptPool.balanceOf(msg.sender)
+        );
+
+        // exit PT position
+        _vault.exitPool(
+            _ptOutInfo.poolId,
+            address(this),
+            payable(address(this)),
+            _ptOutInfo.request
+        );
+
+        // withdraw principal here. This should accumulate underlying.
+        trancheFrom.withdrawPrincipal(
+            trancheFrom.balanceOf(address(this)),
+            address(this)
+        );
+
+        if (!_onlyPrincipal) {
+            // transfer yt lp tokens
+            ytPool = IERC20(
+                address(uint160(uint256(_ytOutInfo.poolId)) >> (12 * 8))
+            );
+            ytPool.transferFrom(
+                msg.sender,
+                address(this),
+                ytPool.balanceOf(msg.sender)
+            );
+
+            // exit YT position
+            _vault.exitPool(
+                _ytOutInfo.poolId,
+                address(this),
+                payable(address(this)),
+                _ytOutInfo.request
+            );
+
+            // withdraw interest to the new tranche to prep a pre-funded deposit
+            trancheFrom.withdrawInterest(
+                IERC20(address(trancheFrom.interestToken())).balanceOf(
+                    address(this)
+                ),
+                address(this)
+            );
+        }
+
+        // transfer required amount and run prefunded deposit
+        trancheFrom.underlying().transfer(
+            address(trancheTo.position()),
+            _toMint
+        );
+        trancheTo.prefundedDeposit(address(this));
+
+        uint256 ptLpBalanceBefore = ptPool.balanceOf(msg.sender);
+        // enter new PT position
+        _vault.joinPool(
+            _ptInInfo.poolId,
+            address(this),
+            payable(msg.sender), // hard-code for security
+            _ptInInfo.request
+        );
+        uint256 ptLpBalanceAfter = ptPool.balanceOf(msg.sender);
+
+        require(
+            ptLpBalanceAfter - ptLpBalanceBefore >= _ptInInfo.lpCheck,
+            "not enough PT LP minted"
+        );
+
+        if (!_onlyPrincipal) {
+            uint256 ytLpBalanceBefore = ytPool.balanceOf(msg.sender);
+            // enter new YT position
+            _vault.joinPool(
+                _ytInInfo.poolId,
+                address(this),
+                payable(msg.sender),
+                _ytInInfo.request
+            );
+            uint256 ytLpBalanceAfter = ptPool.balanceOf(msg.sender);
+            require(
+                ytLpBalanceAfter - ytLpBalanceBefore >= _ytInInfo.lpCheck,
+                "not enough PT LP minted"
+            );
+        }
+
+        // recover any dust
+        IERC20 yt = IERC20(address(trancheFrom.interestToken()));
+        IERC20 t1 = IERC20(address(_ptOutInfo.request.assets[0]));
+        IERC20 t2 = IERC20(address(_ptOutInfo.request.assets[1]));
+
+        yt.transfer(msg.sender, yt.balanceOf(address(this)));
```

Comment:
We should do zeroness checks before transferring dust

Reply:
Possibly go as far as not doing transfers on 'dust amounts', but it's hard to say what they are.